src/BuiltInTools/dotnet-watch/HotReload/BlazorWebAssemblyDeltaApplier.cs

```diff
@@ -31,109 +39,31 @@ public override async Task WaitForProcessRunningAsync(CancellationToken cancella
             // Alternatively, we could inject agent into blazor-devserver.dll and establish a connection on the named pipe.
             => await browserRefreshServer.WaitForClientConnectionAsync(cancellationToken);
 
-        public override async Task<ImmutableArray<string>> GetApplyUpdateCapabilitiesAsync(CancellationToken cancellationToken)
+        public override Task<ImmutableArray<string>> GetApplyUpdateCapabilitiesAsync(CancellationToken cancellationToken)
         {
-            var cachedCapabilities = _cachedCapabilities;
-            if (!cachedCapabilities.IsDefault)
-            {
-                return cachedCapabilities;
-            }
+            var capabilities = project.GetWebAssemblyCapabilities();
 
-            await _capabilityRetrievalSemaphore.WaitAsync(cancellationToken);
-            try
-            {
-                if (_cachedCapabilities.IsDefault)
-                {
-                    _cachedCapabilities = await RetrieveAsync(cancellationToken);
-                }
-            }
-            finally
+            if (capabilities.IsEmpty)
             {
-                _capabilityRetrievalSemaphore.Release();
-            }
+                var targetFramework = project.GetTargetFrameworkVersion();
 
-            return _cachedCapabilities;
-
-            async Task<ImmutableArray<string>> RetrieveAsync(CancellationToken cancellationToken)
-            {
-                var buffer = ArrayPool<byte>.Shared.Rent(32 * 1024);
+                Reporter.Verbose($"Using capabilities based on target framework: '{targetFramework}'.");
 
-                try
+                capabilities = targetFramework?.Major switch
                 {
-                    Reporter.Verbose("Connecting to the browser.");
-
-                    await browserRefreshServer.WaitForClientConnectionAsync(cancellationToken);
-
-                    string capabilities;
-                    if (browserRefreshServer.Options.TestFlags.HasFlag(TestFlags.MockBrowser))
-                    {
-                        // When testing return default capabilities without connecting to an actual browser.
-                        capabilities = GetDefaultCapabilities(targetFrameworkVersion);
-                    }
-                    else
-                    {
-                        string? capabilityString = null;
-
-                        await browserRefreshServer.SendAndReceiveAsync(
```

Comment:
Here's my read: In the past, the runtime would add a new hot reload capability, and we would detect it with this code; there was a fallback based on TFM alone if the runtime was unreachable.  In the future, the project (has access to SDK/build targets/build props) has the main info, and there is a fallback based on TFM alone in case the SDK/build targets don't return everything.
Does this sound right?

Reply:
Yes, the TFM based fallback is for projects referencing older packages that do not set the msbuild property.